### ISSUE TITLE
Discussion: should WebClient use HttpClient as a backend?

### ISSUE THREAD

**Issue by migueldeicaza** [mono/mono#8252]
Currently out WebClient “simple api” is implemented on top of HttpWebRequest, which comes with all the downsides of it - limited HTTP support, not able to consume system settings and in iOS is unable to continue working even if the application is suspended.

Given that this is a high level api perhaps we can use HttpClient, even if we need to load the assembly dynamically or move the actual client into System.

**Comment by filipnavara** [comment]
We mix WebClient and HttpWebRequest in our application and it would be weird if they started behaving differently, even in small ways.

I am not sure what you mean by limited HTTP support. Honestly, the WebClient API is so limited that it shouldn't really matter. Neither do I understand the bit about consuming system settings. WebClient and HttpWebRequest use system proxy settings and the system TLS stack. Is there anything else?

The last thing you mention is application suspend. If that really needs HttpClient, and if using HttpClient directly is not sufficient, then it could easily be supported by a small library for iOS developers that creates a WebClient-like class on top of HttpClient.

**Comment by akoeplinger** [comment]
WebClient on .NET Core actually already does this, i.e. it is reimplemented on top of HttpClient.

**Comment by filipnavara** [comment]
Then again, .NET Core also implements `HttpWebRequest` on top of System.Net.Http, so it is consistent.

**Comment by filipnavara** [comment]
@akoeplinger I don't think you are wrong. It is implemented on top of WebRequest/WebResponse, but those are in turn implemented using System.Net.Http.

I didn't notice it before and it's certainly an approach that may be worth considering. It would keep the stack consistent at all levels. I can see how it could possibly be problematic for custom implementations of HttpMessageHandler implemented on top of HttpWebRequest, but there's a change that it would work.

**Comment by marek-safar** [comment]
That's correct .NET Core wraps HttpWebRequest around HttpClient making it transparent to the end users using WebClient.

I think this is rarely used API which is not worth touching. I am even thinking of making it Obsolete for Xamarin profiles to encourage anyone still using it to move to HttpClient.

**Comment by migueldeicaza** [comment]
Marking it obsolete is really out of the question.

It is an API that is a lot simpler to use than HttpClient.   Let us keep this focused on the pros/cons on replacing the transport for WebClient.

**Comment by marek-safar** [comment]
Is the API really simpler to use?

WebClient:
```csharp
	byte[] data = new WebClient ().DownloadData ("some-url");
```

HttpClient
```csharp
	byte[] data = new HttpClient ().GetByteArrayAsync ("some-url").Result;
```

This one simple use case. I could probably construct special case where WebClient is litle bit shorted but as general API I don't think the risk of breaking possibly anyone is worth taking

**Comment by baulig** [comment]
We have switched to using the WebClient implementation from CoreFX a while ago and I don't really want to switch back to a Mono-specific version, so we should really discuss that with the CoreFX team.
